cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,627 @@
+        -- some other reason, we'll take advantage of that.
+    
+
+  where
+    probeMergeResult :: MergeResult (FilePath, FileGlobMonitorState) FilePath
+                     -> ChangedM (FilePath, FileGlobMonitorState)
+
+    -- Only in cached (directory deleted)
+    probeMergeResult (OnlyInLeft (path, cgp))
+      | not (hasMatchingFiles cgp) = return (path, cgp)
+        -- Strictly speaking we should be returning 'CacheChanged' above
+        -- as we should prune the now-missing 'FileGlobMonitorState'. However
+        -- we currently just leave these now-redundant entries in the
+        -- cache as they cost no IO and keeping them allows us to avoid
+        -- rewriting the cache.
+      | otherwise = somethingChanged path
+
+    -- Only in current filesystem state (directory added)
+    probeMergeResult (OnlyInRight path) = do
+      cgp <- liftIO $ buildFileGlobMonitorState root (dirName </> path) globPath
+      if hasMatchingFiles cgp
+        then somethingChanged path
+        else cacheChanged >> return (path, cgp)
+
+    -- Found in path
+    probeMergeResult (InBoth (path, cgp) _) = do
+      cgp' <- probeGlobStatus root (dirName </> path) cgp
+      return (path, cgp')
+
+    -- | Does a 'FileGlobMonitorState' have any relevant files within it?
+    hasMatchingFiles :: FileGlobMonitorState -> Bool
+    hasMatchingFiles (MonitorStateGlobFiles _ _   entries) = not (null entries)
+    hasMatchingFiles (MonitorStateGlobDirs  _ _ _ entries) =
+      any (hasMatchingFiles . snd) entries
+
+
+probeGlobStatus root dirName (MonitorStateGlobFiles glob mtime children) = do
+    change <- liftIO $ checkDirectoryModificationTime (root </> dirName) mtime
+    mtime' <- case change of
+      Nothing     -> return mtime
+      Just mtime' -> do
+        -- directory modification time changed:
+        -- a matching file may have been added or deleted
+        matches <- filterM (\entry -> let file = root </> dirName </> entry
+                                       in liftIO $ doesFileExist file)
+                 . filter (globMatches glob)
+               =<< liftIO (getDirectoryContents (root </> dirName))
+
+        let mergeRes = mergeBy (\(path1,_,_) path2 -> compare path1 path2)
+                         children
+                         (sort matches)
+        unless (all isInBoth mergeRes) (somethingChanged dirName)
+        return mtime'
+
+    -- Check that none of the children have changed
+    forM_ children $ \(file, fmtime, fhash) ->
+        probeFileModificationTimeAndHash root (dirName </> file) fmtime fhash
+
+    return (MonitorStateGlobFiles glob mtime' children)
+    -- Again, we don't force a cache rewite with 'cacheChanged', but we do use
+    -- the new mtime' if any.
+  where
+    isInBoth :: MergeResult a b -> Bool
+    isInBoth (InBoth _ _) = True
+    isInBoth _            = False
+
+------------------------------------------------------------------------------
```

Comment:
I don't know what any of this does and I'd really appreciate short pointers on what these are supposed to do.